bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java

```diff
@@ -292,9 +294,12 @@ public void widgetSelected(SelectionEvent e) {
 						setupFindReplaceLogic();
 						boolean eventRequiresInverseSearchDirection = (e.stateMask & SWT.MODIFIER_MASK) == SWT.SHIFT;
 						boolean forwardSearchActivated = findReplaceLogic.isActive(SearchOptions.FORWARD);
+						boolean incrementalSearchActivated = findReplaceLogic.isActive(SearchOptions.INCREMENTAL);
 						activateInFindReplaceLogicIf(SearchOptions.FORWARD,
 								eventRequiresInverseSearchDirection != forwardSearchActivated);
+						findReplaceLogic.deactivate(SearchOptions.INCREMENTAL);
 						boolean somethingFound = findReplaceLogic.performSearch(getFindString());
+						activateInFindReplaceLogicIf(SearchOptions.INCREMENTAL, incrementalSearchActivated);
```

Comment:
Why is it necessary to deactive incremental search while performing an explicit search operation? To me this looks like a workaround, because the logic does not behave as xpected.

Reply:
The specification dictates the inverse (i.e. the old implementation was wrong): if `INCREMENTAL` is active, then every search should be incremental unless incremental is impossible (we didn't specify any exceptions, except for that case)

In this case, this makes sense: we temporarily disable the incremental mode in the Logic (i.e. we want to perform a normal search) and re-enable it. Of course, we could add a helper method for that in the Logic but that would only render the interface less coherent (because suddenly we introduce a case where a search option can sometimes be disregarded by the Logic)

Reply:
Okay, I think there is/was a slight misunderstanding about the meaning of "incremental" on my side. So actually, the "incremental" option in the UI does not totally activate incremental search (i.e., the option in the logic), but it activates some "incremental search-as-you-type". With that in mind, the change makes sense. I have to admit that for me it is still hard to comprehend how the interaction between incremental and non-incremental search operations works or is supposed to work (as incremental mode only seems to be used for search-as-you-type). Maybe we can once find an even better design to reflect the behaviors. For now, having the logic in either incremental or non-incremental mode and only a single `performSearch` operation is good to make it easier to understand and use. So if you are confident that this change is semantics-preserving, it is good to go from my side.

Reply:
>  So actually, the "incremental" option in the UI does not totally activate incremental search (i.e., the option in the logic), but it activates some "incremental search-as-you-type"

UI-Wise what happens is that indeed every input to the Editor will perform a "search as you type".
However yes, when the user presses the "search" button of this dialog, he expects to get to the next search result. The logic should not care about this as this is a pure UI decision (some UIs might, for example, not do a "search as you type" and instead use a search button to trigger the incremental search).

An alternative would be to not store Incremental search as a search option but to instead have another method which just runs the incremental search. But then, the Dialog would need to store that it is set to run Incremental search itself, which would split the concern of storing parametrization over two classes.

Reply:
> However yes, when the user presses the "search" button of this dialog, he expects to get to the next search result.

Exactly. So the "incremental" option in the dialog is actually not an option for the search overall but activates search-as-you-type for the input box (including incremental search execution when doing search-as-you-type). That's might be confusing (but is not related to your changes but has just always been like this).